### PR TITLE
libs/libc/string: add memmem

### PR DESCRIPTION
**Pull request by Cynerd**
This function is the GNU extension.

## Testing

Tested on custom SAME70N21 based board.

### CODE REVIEW

File: libs/libc/string/lib_memmem.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
```
Comment (Donny9) [inline libs/libc/string/lib_memmem.c:25]: remove

File: libs/libc/string/lib_memmem.c [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ *   The memmem() function returns a pointer to the beginning of the
+ *   substring, or NULL if the substring is not found.
+ *
+ ****************************************************************************/
+
+FAR void  *memmem(FAR const void *haystack, size_t haystacklen,
+                  FAR const void *needle, size_t needlelen)
+{
+  size_t i, y;
+  const unsigned char *h = haystack;
```
Comment (Donny9) [inline libs/libc/string/lib_memmem.c:51]: add FAR,ditto

File: libs/libc/string/lib_memmem.c [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  const unsigned char *n = needle;
+
+  if (needlelen > haystacklen)
+    return NULL;
+
+  for (i = 0; i < (haystacklen - needlelen); i++)
+    {
+      y = 0;
+      while (h[i + y] == n[y])
+      {
```
Comment (Donny9) [inline libs/libc/string/lib_memmem.c:61]: run checkpatch.sh to check

File: libs/libc/string/lib_memmem.c (hunk 2 above)
Comment (Cynerd) [reply libs/libc/string/lib_memmem.c]: Added... good point.

File: libs/libc/string/lib_memmem.c (hunk 1 above)
Comment (Cynerd) [reply libs/libc/string/lib_memmem.c]: I copied it from other files, but it is true that it is not necessary here at all.

File: libs/libc/string/lib_memmem.c [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ *   substring needle of length needlelen in the memory area haystack of
+ *   length haystacklen.
+ *
+ * Returned Value:
+ *   The memmem() function returns a pointer to the beginning of the
+ *   substring, or NULL if the substring is not found.
+ *
+ ****************************************************************************/
+
+FAR void  *memmem(FAR const void *haystack, size_t haystacklen,
```
Comment (xiaoxiang781216) [inline libs/libc/string/lib_memmem.c:45]: remove extra space before *

File: libs/libc/string/lib_memmem.c [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+                  FAR const void *needle, size_t needlelen)
+{
+  size_t i;
+  size_t y;
```
Comment (xiaoxiang781216) [inline libs/libc/string/lib_memmem.c:49]: move i and y after line 51?

File: libs/libc/string/lib_memmem.c [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  FAR const unsigned char *h = haystack;
+  FAR const unsigned char *n = needle;
+
+  if (needlelen > haystacklen)
+    {
+      return NULL;
+    }
+
+  for (i = 0; i < (haystacklen - needlelen); i++)
```
Comment (xiaoxiang781216) [inline libs/libc/string/lib_memmem.c:58]: remove () around haystacklen - needlelen?

**Comment by xiaoxiang781216** [comment]
@Cynerd let's squash these patch into one.

File: libs/libc/string/lib_memmem.c (hunk 6 above)
Comment (Cynerd) [reply libs/libc/string/lib_memmem.c]: Honestly, I am used to adding them when math gets long but not necessary here for sure.

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Comment by pkarashchenko** [comment]
Please fix build error
```
string/lib_memmem.c: In function 'memmem':
Error: string/lib_memmem.c:65:24: error: return discards 'const' qualifier from pointer target type [-Werror=discarded-qualifiers]
   65 |               return h + i;
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**